presto-main/src/main/java/com/facebook/presto/sql/MaterializedViewUtils.java

```diff
@@ -0,0 +1,88 @@
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.metadata.MetadataUtil.toSchemaTableName;
+import static com.facebook.presto.spi.ConnectorMaterializedViewDefinition.TableColumn;
+import static com.facebook.presto.sql.analyzer.MaterializedViewPlanValidator.MaterializedViewPlanValidatorContext;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+
+public class MaterializedViewUtils
+{
+    private MaterializedViewUtils() {}
+
+    public static void validateMaterializedViewQueryPlan(Statement query)
+    {
+        MaterializedViewPlanValidator validator = new MaterializedViewPlanValidator(query);
+        validator.process(query, new MaterializedViewPlanValidatorContext());
+    }
+
+    public static Map<String, Map<SchemaTableName, String>> computeMaterializedViewToBaseTableColumnMappings(Query viewQuery, Analysis analysis)
+    {
+        ImmutableMap.Builder<String, Map<SchemaTableName, String>> fullColumnMapping = ImmutableMap.builder();
+
+        Map<String, Map<SchemaTableName, String>> originalColumnMapping = getOriginalColumnMappingFromAnalysis(analysis, viewQuery);
+
+        MaterializedViewPlanValidator planAnalyzer = new MaterializedViewPlanValidator(viewQuery, true, Optional.of(analysis));
+        planAnalyzer.process(viewQuery, new MaterializedViewPlanValidatorContext());
+        List<List<TableColumn>> linkedBaseColumns = planAnalyzer.getLinkedBaseColumns();
+
+        for (Map.Entry<String, Map<SchemaTableName, String>> entry : originalColumnMapping.entrySet()) {
+            String viewColumn = entry.getKey();
+            Map<SchemaTableName, String> originalBaseColumns = entry.getValue();
+
+            Map<SchemaTableName, String> fullBaseColumns = new HashMap<SchemaTableName, String>() {{
+                    putAll(originalBaseColumns); }};
+
+            originalBaseColumns.forEach((originalBaseTable, originalBaseColumn) -> {
+                TableColumn originalTableColumn = new TableColumn(originalBaseTable, originalBaseColumn);
+                linkedBaseColumns.forEach(linkedTableColumnPair -> {
+                    if (originalTableColumn.equals(linkedTableColumnPair.get(0))) {
+                        fullBaseColumns.put(linkedTableColumnPair.get(1).getTableName(), linkedTableColumnPair.get(1).getColumnName());
+                    }
+                    else if (originalTableColumn.equals(linkedTableColumnPair.get(1))) {
+                        fullBaseColumns.put(linkedTableColumnPair.get(0).getTableName(), linkedTableColumnPair.get(0).getColumnName());
+                    }
+                });
+            });
+
+            fullColumnMapping.put(viewColumn, ImmutableMap.copyOf(fullBaseColumns));
+        }
+
+        return fullColumnMapping.build();
+    }
+
+    private static Map<String, Map<SchemaTableName, String>> getOriginalColumnMappingFromAnalysis(Analysis analysis, Node node)
```

Comment:
s/node/view. Also, `Node` is too generic. I assume we are talking about `Query` here.

Reply:
It's actually a generic func that works for any Node. Since now we place it inside MaterializedViewUtils, I will name it viewQuery.